core/blockchain_snapshot_test.go

```diff
@@ -0,0 +1,469 @@
+// In this case the chain head should be rewound to the point
+// with available state. And also the new head should must be lower
+// than disk layer. But there is only a low committed point so the
+// chain should be rewound to committed point.
+func TestLowCommitCrashWithNewSnapshot(t *testing.T) {
+	testSnapshot(t, &snapshotTest{
+		legacy:             false,
+		crash:              true,
+		gapped:             0,
+		setHead:            0,
+		chainBlocks:        8,
+		snapshotBlock:      4,
+		commitBlock:        2,
+		expCanonicalBlocks: 8,
+		expHeadHeader:      8,
+		expHeadFastBlock:   8,
+		expHeadBlock:       2,
+		expSnapshotBottom:  4, // Last committed disk layer, wait recovery
+	})
+}
+
+// Tests a Geth was crashed and restarts with a broken snapshot.
+// In this case the chain head should be rewound to the point
+// with available state. And also the new head should must be lower
+// than disk layer. But there is only a high committed point so the
+// chain should be rewound to genesis.
+func TestHighCommitCrashWithNewSnapshot(t *testing.T) {
+	testSnapshot(t, &snapshotTest{
+		legacy:             false,
+		crash:              true,
+		gapped:             0,
+		setHead:            0,
+		chainBlocks:        8,
+		snapshotBlock:      4,
+		commitBlock:        6,
+		expCanonicalBlocks: 8,
+		expHeadHeader:      8,
+		expHeadFastBlock:   8,
+		expHeadBlock:       0,
+		expSnapshotBottom:  4, // Last committed disk layer, wait recovery
+	})
+}
+
+// Tests a Geth was crashed and restarts with a broken and "legacy format"
+// snapshot. In this case the entire legacy snapshot should be discared
+// and rebuild from the new chain head.
+func TestNoCommitCrashWithLegacySnapshot(t *testing.T) {
+	testSnapshot(t, &snapshotTest{
+		legacy:             true,
+		crash:              true,
+		gapped:             0,
+		setHead:            0,
+		chainBlocks:        8,
+		snapshotBlock:      4,
+		commitBlock:        0,
+		expCanonicalBlocks: 8,
+		expHeadHeader:      8,
+		expHeadFastBlock:   8,
+		expHeadBlock:       0,
+		expSnapshotBottom:  4, // Last committed disk layer, wait recovery
+	})
+}
+
+// Tests a Geth was crashed and restarts with a broken and "legacy format"
+// snapshot. In this case the entire legacy snapshot should be discared
+// and rebuild from the new chain head.
+//
+// The special thing in this test case is: we only persist the snasphot
+// journal during the shutdown previously. In this case after the crash
+// there is no journal at all.
+func TestLowCommitCrashWithLegacySnapshot(t *testing.T) {
+	testSnapshot(t, &snapshotTest{
+		legacy:             true,
+		crash:              true,
+		gapped:             0,
+		setHead:            0,
+		chainBlocks:        8,
+		snapshotBlock:      4,
+		commitBlock:        2,
+		expCanonicalBlocks: 8,
+		expHeadHeader:      8,
+		expHeadFastBlock:   8,
+		expHeadBlock:       2,
+		expSnapshotBottom:  4, // Last committed disk layer, wait recovery
```

Comment:
Same issue as above, tests says we should regen from head, but the exp bottom in the scenario says otherwise.

```
Chain:
  G->C1->C2->C3->C4->C5->C6->C7->C8 (HEAD)

Commit:   G, C2
Snapshot: G, C4

CRASH

------------------------------

Expected in leveldb:
  G->C1->C2->C3->C4->C5->C6->C7->C8
Expected head header    : C8
Expected head fast block: C8
Expected head block     : C2
Expected snapshot disk  : C4
```